core/src/main/scala/besom/util/NonEmptyString.scala

```diff
@@ -39,8 +64,34 @@ object NonEmptyString:
   private def fromImpl(expr: Expr[String])(using quotes: Quotes): Expr[NonEmptyString] =
     import quotes.reflect.*
 
+    // kudos to @Kordyjan for most of this implementation
+    def downTheRabbitHole(tree: Term): Either[Term, String] =
```

Comment:
this is quire dense, I would appreciate adding some comments to guide the reader through this code, esp. for people not fluent in macros

I understand that we are recursively examining the AST and trying to figure out all cases where it can be proven the result string will be not empty, but I'm sure I'm missing some important details.

In the interest of maintenance, adding comments would be greatly appreciated.